scipio/src/controllers/deadline_queue.rs

```diff
@@ -0,0 +1,470 @@
+///   file: you may be moving data to the file internal buffers but they are not *flushed* to the
+///   media. When the data is finally flushed a giant bubble is inserted into the control loop. The
+///   characteristics of the system will radically change. Contract that for instance with O_DIRECT
+///   files, where writing to the file means writing to the media: smooth and fast feedback!
+///
+///   The moral of the story is:
+///    * do not use this with buffered files or other buffered sinks where where the real physical response
+///      is delayed
+///    * do not set the adjustment period too low
+///    * do not use this very short lived processes.
+///
+/// To calculate the speed of the process, the needs of all elements in the queue are considered.
+///
+/// Let's say, for instance, that you queue items A, B and C, each with 10,000 units finishing
+/// respectively in 1, 2 and 3 minutes. From the point of view of the system, all units from A
+/// and B need to be flushed before C can start so that is taken into account: the system needs
+/// to set its speed to 10,000 points per minute so that the entire queue is flushed in 3
+/// minutes.
+pub struct DeadlineQueue<T> {
+    tq: TaskQueueHandle,
+    sender: LocalSender<Rc<dyn DeadlineSource<Output = T>>>,
+    handle: task::join_handle::JoinHandle<(), ()>,
+    queue: Rc<InnerQueue<T>>,
+}
+
+impl<T: 'static> DeadlineQueue<T> {
+    /// Creates a new `DeadlineQueue` with a given `name` and `adjustment period`
+    ///
+    /// Internally the `DeadlineQueue` spawns a new task queue with dynamic shares
+    /// in which it will execute the tasks that were registered.
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::LocalExecutor;
+    /// use scipio::controllers::DeadlineQueue;
+    /// use std::time::Duration;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    ///
+    /// ex.run(async {
+    ///     DeadlineQueue::<usize>::new("example", Duration::from_millis(250));
+    /// });
+    /// ```
+    pub fn new(name: &'static str, adjustment_period: Duration) -> DeadlineQueue<T> {
+        let queue = Rc::new(InnerQueue::new(adjustment_period));
+        // We could always dispatch into the latency queue, but since that effectively means
+        // putting requests in a different io_uring we'll avoid doing that unless the process
+        // is very sensitive. Because of dead time it wouldn't be a bad idea to even have a minimum
+        // here. But we'll just document it and leave it to the user.
+        let lat = {
+            if adjustment_period < Duration::from_millis(100) {
+                Latency::Matters(adjustment_period)
+            } else {
+                Latency::NotImportant
+            }
+        };
+
+        let tq = Local::create_task_queue(Shares::Dynamic(queue.clone()), lat, name);
+        let (sender, receiver): (LocalSender<QueueItem<T>>, LocalReceiver<QueueItem<T>>) =
+            LocalChannel::new_bounded(1);
```

Comment:
channels APIs usually make the ctor function top-level. That way, you'll have a nice place to put a type parameter once: `local_channel::new_bounded::<QueueItem<T>>(1)`

Reply:
so just a free function without an associated struct?
I'll change that in the channel PR.